### PR TITLE
Statistics.koplugin: calendar day detail

### PR DESCRIPTION
**Pull request by weijiuqiao**
This PR adds a detail view of calendar day for the statistics plugin.

image

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9813)

### CODE REVIEW

**Comment by poire-z** [comment]
That's awesomely neat !

And quite revealing.... I'm a fragmented reader :/

That's mostly 2 hours of reading:
image

and that was 5 hours of serious and passionate (!) reading of `Chinese typography - The Type`:
image

I trust the stats, and the reasons for these are either:
- My "Read page duration limit: 20s / 240s" are badly tuned, so the holes, each time I may spend 250s on a page.
- (I have some personal code to handle "tap bottom corner to standby" that I may use more often than people who suspend and need the power button to resume - and that I must abuse more than I thought...)

So, I don't know if other people will see such fragmentation, with the book titles so often cut - and if that is just fine (showing the harsh reality :), or if you should/could ensure/simulate some continuation when the gap is small.

Some other remarks (you may already have some of these in mind):
- the split to multiple pages when there are >= 6 books on a day is a bit cumbersome, acting like KVPage on the whole thing with prev/next arrows at the bottom, while the timeline sticks. Having the arrows under the top section would make this less odd, but well, may be hard to do and not so practical (having to move fingers up there) - and 6 books per days is quite rare - I usually get 2,3,4 and rarely 5, that all generates only a single page.
- with a single page (or even if multiple page, when reaching first/last), I would expect a swipe east/west to show me the prev or next day. Or/and may be north/south swipes too, as the timeline rolls from top 0h00 to bottom 23h :)
- May be the titles in the book rectangles in the top section could have a little more left padding (and so, right too)
- The checkboxes are fun to play with :)

File: plugins/statistics.koplugin/calendarview.lua [hunk 1]
```diff
@@ -719,11 +1237,18 @@ function CalendarView:_populateItems()
...
             callback = not is_future and function()
-                -- Just as ReaderStatistics:callbackDaily(), but without any window stacking
-                UIManager:show(KeyValuePage:new{
+                local kv_pairs = self.reader_statistics:getBooksFromPeriod(day_ts, day_ts + 86400)
+                local seconds_books = self.reader_statistics:getReadingDurationBySecond(day_s)
+                for _, kv in ipairs(kv_pairs) do
+                    if seconds_books[kv.book_id] then
+                        kv.periods = seconds_books[kv.book_id].periods
+                    end
+                    kv.checked = true
```
Comment (poire-z) [inline plugins/statistics.koplugin/calendarview.lua:1246]: Could it happen some books get to be unchecked ?

File: plugins/statistics.koplugin/main.lua [hunk 2]
```diff
@@ -2687,6 +2688,43 @@ function ReaderStatistics:getReadBookByDay(month)
...
 end
 
+function ReaderStatistics:getReadingDurationBySecond(date)
+    local sql_stmt = [[
+        SELECT
+            start_time - strftime('%s', ?, 'utc') as start,
+            start_time - strftime('%s', ?, 'utc') + duration as end,
+            id_book book_id,
+            book.title book_title
+        FROM   page_stat_data
```
Comment (poire-z) [inline plugins/statistics.koplugin/main.lua:2698]: By using `page_stat_data`, I guess we avoid some computations done by the `page_stat` "view" (unless sqlite is clever enough to figure it's not needed to use it and it can just access the real table underneath).
May be some other SQL queries could benefit from also using `page_stat_data` ? (It's possible we/I made these SQL before the page_stat_data / page_stat view were introduced).

File: plugins/statistics.koplugin/main.lua [hunk 3]
```diff
@@ -2687,6 +2688,43 @@ function ReaderStatistics:getReadBookByDay(month)
...
+        JOIN   book ON book.id = page_stat_data.id_book
+        WHERE  start_time BETWEEN strftime('%s', ?, 'utc')
+                              AND strftime('%s', ?, 'utc', '+24 hours', '-1 second')
```
Comment (poire-z) [inline plugins/statistics.koplugin/main.lua:2701]: Any idea why we/I used 'utc' in similar queries ? There must be a good reason, but I don't remember/see why :)
May be because we insert without more thinking needed as utc times, and it's just fine on our devices. But we sometimes use 'localtime' when formatting - all this may just work and is right - just asking if you wondered and give it some thoughts, and if that makes sense to you :)

File: plugins/statistics.koplugin/main.lua [hunk 4]
```diff
@@ -1942,6 +1942,7 @@ function ReaderStatistics:getBooksFromPeriod(period_begin, period_end, callback_
         table.insert(results, {
             result_book[1][i],
             T(_("%1 (%2)"), util.secondsToClockDuration(user_duration_format, tonumber(result_book[2][i]), false), tonumber(result_book[3][i])),
+            book_id = tonumber(result_book[4][i]),
```
Comment (poire-z) [inline plugins/statistics.koplugin/main.lua:1945]: Why was that needed ? Or it is now with your added code ?

File: plugins/statistics.koplugin/calendarview.lua [hunk 5]
```diff
@@ -364,6 +365,523 @@ function CalendarWeek:update()
...
+        UIManager:setDirty(self, function()
+            return "ui", self.check_container.dimen
+        end)
+    elseif self.item.callback then
+        self.item:callback()
+    end
+    return true
+end
+
+local CalendarDayDetail = FocusManager:extend{
```
Comment (poire-z) [inline plugins/statistics.koplugin/calendarview.lua:483]: May be name this `CalenderDayView` ? to make it clearer it's a fullscreen widget like CalendarView and not a sub-component widget?

**Review by poire-z — COMMENTED** [review]
Impressive how you came up with that in 1 or 2 days - would have taken me 1 or 2 weeks :)

File: plugins/statistics.koplugin/calendarview.lua (hunk 1 above)
Comment (weijiuqiao) [reply plugins/statistics.koplugin/calendarview.lua:1246]: I had thought about only checking the top couple ones but it turned out for me that the others have barely any bars so turning them on/off doesn't make much difference.

File: plugins/statistics.koplugin/main.lua (hunk 2 above)
Comment (weijiuqiao) [reply plugins/statistics.koplugin/main.lua:2698]: >unless sqlite is clever enough to figure

I doubt that, but I'm not expert in the inner workings of sqlite...

>May be some other SQL queries could benefit from also using page_stat_data ?

I did a quick check and many of them use the `page` column, on which the page_stat view seem to be the better choice :)

File: plugins/statistics.koplugin/main.lua (hunk 3 above)
Comment (weijiuqiao) [reply plugins/statistics.koplugin/main.lua:2701]: I copied them from other queries :) Guess it's because the timestamps we get from `os.time()` and inserted to db are utc-based? Here the `?`s are bound with local time strings like '2022-11-11', so it has to be converted to utc seconds?

File: plugins/statistics.koplugin/main.lua (hunk 4 above)
Comment (weijiuqiao) [reply plugins/statistics.koplugin/main.lua:1945]: I need this to pair book with the reading periods that I got from the other query, otherwise only title can be used and there might be duplicate titles for different book entries.

**Comment by weijiuqiao** [comment]
>or if you should/could ensure/simulate some continuation when the gap is small

I'm also considering about it. How small is a little hard to decide though. Changed to merge the gap if smaller than the minimum seconds set by user.

> Having the arrows under the top section would make this less odd, but well, may be hard to do and not so practical (having to move fingers up there) - and 6 books per days is quite rare - I usually get 2,3,4 and rarely 5, that all generates only a single page.

As you said, and it might look a little odd with the arrows in the middle. But some menus do this, then a line to separate the two sections will be needed. It's not super hard to achieve though, just pending the page_info to the book_items vertical group widget.

>with a single page (or even if multiple page, when reaching first/last), I would expect a swipe east/west to show me the prev or next day. Or/and may be north/south swipes too, as the timeline rolls from top 0h00 to bottom 23h :)

Makes sense. But there'll be another caveat: when we swipe to another month do we return to the previous month calendar view or the new one?

>May be the titles in the book rectangles in the top section could have a little more left padding (and so, right too)

Done.

>The checkboxes are fun to play with :)

Good to know :)

File: plugins/statistics.koplugin/main.lua (hunk 3 above)
Comment (weijiuqiao) [reply plugins/statistics.koplugin/main.lua]: Turns out I can just use the timestamp instead of date strings. Changed to use timestamp.

**Comment by weijiuqiao** [comment]
Added swipe left/right to go the prev/next day, but still return to the old month.

File: plugins/statistics.koplugin/main.lua (hunk 4 above)
Comment (poire-z) [reply plugins/statistics.koplugin/main.lua:1945]: Oups, my mind had read that single line as:
```diff
-book_id = result_book[4][i],
+book_id = tonumber(result_book[4][i]),
```
so my questions were about this imaginary diff :)

**Comment by poire-z** [comment]
> Changed to merge the gap if smaller than the minimum seconds set by user.

I may get a little less fragmentation, but still some - which might be just fine.

> But there'll be another caveat: when we swipe to another month do we return to the previous month calendar view or the new one?
> Added swipe left/right to go the prev/next day, but still return to the old month.

Dunno. It's indeed hard to visually know where you were when you return from day to month. Even when not swiping days. So, somehow, I guess that I would be fine if I would return to the new month. (Should be achievable with a on_close_callback on the dayView, refreshing the month view?)

Minor aesthetic remark: I wonder if these edges could/should be aligned:
image

Also, with the variable hourly sections heights, depending on the nb of book titles that day, when swiping days, and separator lines not being drawn over the preivous ones, the screen on my old GloHD gets quickly dirty, with the ghosting of the old lines. May be it won't happen on newer devices, and I don't think we should trigger flashing refreshes just for that. (This would not happen if the hourly lines had a fixed width, but then it would not look as filled/full/nice.
(Note that if we "fix" my "Minor aesthetic remark" above, we'll get some ghosting on the variabling possitionned book titles.)

**Comment by weijiuqiao** [comment]
Widened gap between checkmark and title to fit in the numbers:

image

Now returns to month based on current day.

**Comment by poire-z** [comment]
> Widened gap between checkmark and title to fit in the numbers:

Looks like you did not, and just made the hours (00:00 01:00...) always having the smallest size, right? Which is fine.

Btw, it looks like you removed the previous feature of long-press on kvpage items (this day's book titles) to remove the stats for that book for that day. Or I can't find it. I'd like to keep it :)

**Comment by weijiuqiao** [comment]
> it looks like you removed the previous feature of long-press on kvpage items (this day's book titles) to remove the stats 

Didn't know it was there. Restored.

**Comment by poire-z** [comment]
More thoughts:
- about the ghosting (I checked that using black lines, instead of gray lines, doesn't make it better), may be when closing the top Calendar view, you should do a flashing refresh? I didn't feel the need for it before - but it looks like I did that it in Book map and Page browsers - so it wouldn't feel odd to have it done when leaving Calendar which also has some big black/gray bars and lines.
- dunno if it's just (noticable) optical illustion (or integer arithmetic desillusion), but it seems the bottom padding is larger than the top padding:
image
- may be just me playing with it as it is new - but it could be nice to have an event in Dispatcher to get the current day CalendarDayView, so we can assign it to a gesture and see our today's progress (and have small challenges like "I won't stop my reading session until I make a bar with the full book title untruncated" :)

**Comment by poire-z** [comment]
Some other small glitches/remarks:
- when swipping days, crossing 2022-10-30, I get this day view displayed twice with that day in the title, but with a different timeline. 2022-10-30 was the day of -1/+1 hour change for daylight saving time here in France, so this might be the reason. I think I solved that situation before (in KOReader or some other project, not sure) by using 03:00 instead of 00:00 as the time reference for a day (so 03:00 - 1 = 02:00 which is still in the targeted day). 
- when leaving day view on another month than the starting month, we get to see the old month for a few ms before the new month view is shown, so some small ugly refresh animation - dunno if this should be avoided, or if this animation can be felt as a hint we indeed changed month and are not on the same as before.

**Comment by weijiuqiao** [comment]
Ghosting removal done.

>it seems the bottom padding is larger than the top padding

Done as good as I can, without using thicker lines.

>but it could be nice to have an event in Dispatcher

Done.

> I think I solved that situation before (in KOReader or some other project, not sure) by using 03:00

Changed to 03:00 for determining date.

> we get to see the old month for a few ms before the new month view is shown, dunno if this should be avoided, or if this animation can be felt as a hint we indeed changed month and are not on the same as before.

This was by design, it was like this before :)

> have small challenges like "I won't stop my reading session until I make a bar with the full book title untruncated"

Interesting idea, another way to encourage oneself to read more :)

File: plugins/statistics.koplugin/calendarview.lua [hunk 6]
```diff
@@ -364,6 +365,609 @@ function CalendarWeek:update()
...
+        })
+
+        if i ~= 0 then
+            table.insert(self.timeline, FrameContainer:new{
+                width = self.timeline_width,
+                height = Size.border.thin,
+                background = Blitbuffer.COLOR_LIGHT_GRAY,
+                bordersize = 0,
+                padding = 0,
+                overlap_offset = { self.time_text_width, offset_y - Size.border.thin/2 },
```
Comment (poire-z) [inline plugins/statistics.koplugin/calendarview.lua:880]: MIght want to add it to Size as `Size.border.thinner` or `Size.border.extra_thin`.
Also, math.floor() if doing that here - we absolutely never want non-integer x/y/w/h, or you're in for a hell of debugging positionning issues :)

Was that your "Done as good as I can, without using thicker lines." ? May be these non-integer values were the cause?
Also, shouldn't you better fix the height of the spans to be `height of the line - separator - 2*some_padding`, to solve thie issue - or then the text is not vertically aligned in the span (althought this might be less noticable?)

**Comment by poire-z** [comment]
- It looks like you reduced the separator thickness - i preferred them before: now, there are so thin that they may not stand out from the ghosted ones :) (also, dunno how we'll do with super thin / math.floored() on low dpi devices, and they may end up invisible?)
- probably another candidate for the 03:00 tweak: when in October month views, and swiping days 2022-10-30 > 2022-10-31 > 2022-11-01, and closing, I get back to October (I expect November). When I swipe to 2022-11-02, and close, I get November. (Seems to not always happen ? Or depending from where I come.)

**Comment by weijiuqiao** [comment]
Well, `Size.border.thin` is 0.5, so half of it is 0.25, both are none-integer.

Also, in the dispatch show day view call, should I do `insertDB()` as in show calendar view? Looks like it will break the reading stats if a new record is not started when I close it and return to the book I'm reading.

File: plugins/statistics.koplugin/main.lua [hunk 7]
```diff
@@ -130,6 +130,7 @@ ReaderStatistics.default_settings = {
 }
 
 function ReaderStatistics:onDispatcherRegisterActions()
+    Dispatcher:registerAction("stats_calendar_day_view", {category="none", event="ShowCalendarDayView", title=_("Open today's statistics timeline"), general=true, separator=false})
```
Comment (poire-z) [inline plugins/statistics.koplugin/main.lua:133]: I spend a few moments looking for it. `Open today's statistics timeline` wasn't catching my eyes as I expected Calendar or statistics as the first word. Also, "Open" is not needed. May be just:
`Statistics today's timeline` ?
Also dunno about the order, most recent added after older ones? Or by increasing granularity?
```
Statistics calendar view
Statistics today's timeline
```

**Comment by poire-z** [comment]
> Well, `Size.border.thin` is 0.5, so half of it is 0.25, both are none-integer.

In size.lua, `thin = Screen:scaleBySize(0.5),` should ensure it's integer, and scaled up. https://github.com/koreader/koreader-base/blob/4bebcda5fdd4cda7e3d3c6092ee21913152fdab4/ffi/framebuffer.lua#L403-L415

> Also, in the dispatch show day view call, should I do `insertDB()` as in show calendar view? Looks like it will break the reading stats if a new record is not started when I close it and return to the book I'm reading.

Dunno :) It's always a bit bothering to get your new page flushed (and possibly ignored if you just browse to it since less than min_duration). But do as best as it is needed to break the less - and too bad for the people who would call this too often.

**Comment by weijiuqiao** [comment]
>In size.lua, thin = Screen:scaleBySize(0.5), should ensure it's integer, and scaled up.

Good to know, since default is (1) and thin (0.5), I'm using them to avoid calling math.floor.
>Dunno :) It's always a bit bothering to get your new page flushed

I removed the flushing, since it's potentially mostly meant for checking if the title has been un-truncated :), it would be too bad to have the bar broken just because you checked.

**Comment by weijiuqiao** [comment]
>I removed the flushing,

Guess I have to flush, otherwise there's no change to see...

**Comment by poire-z** [comment]
> Guess I have to flush, otherwise there's no change to see...

Yep, that's what I assumed would be happening :/
I had the same kind of problem I think on suspend/standby, that I may have fixed in #8745. It feels we could just call/emit onReaderPaused / onReadingResumed when showing/closing these widgets ? (and it that works, also use them when showing the monthly view?

**Comment by weijiuqiao** [comment]
> if I wrote update the page start ts by the time spent sleeping, which is not what we want here :/

Right, it's opposite to what we what here. OnPaused/resume will have the resumed ts as the starting ts, effectively break the streak :)

It turns out as long as we doesn't go beyond the max_sec or have multiple pages under min_sec so the gap added together gets longer that a min_sec, the bar is intact.
This makes me think if we could set a relatively large minimum value before the bars are split. Say max(20, user_set_min_sec) seconds.

**Comment by weijiuqiao** [comment]
Decided on 30s: If there's a gap within max(30s, min_sec) of the same book and there's no other book in-between, then the bar will not break.

**Comment by poire-z** [comment]
I thought you would have gone figuring a way to insertDB without losing the current page/duration :)
What happens if I have min_sec = 30, go to a new page, read 16s, open timeline, close timeline, read another 16s, and then turn page ? A gap ?
But anyway, ok with your trick.

**Comment by weijiuqiao** [comment]
>What happens if I have min_sec = 30, go to a new page, read 16s, open timeline, close timeline, read another 16s, and then turn page ? A gap ?

It seems when inserting db, the current page's data is not inserted, so the page will have 16 + 16 + time_spend_in_timeline duration and no gap?

**Comment by poire-z** [comment]
OK, didn't really check. So, what was the issue precisely that needed your latest commit ?

**Comment by weijiuqiao** [comment]
> what was the issue precisely that needed your latest commit ?

Aesthetics, less fragmentation. Allow users to skim many pages (within 30s) without making a gap.

**Comment by poire-z** [comment]
> Also, in the dispatch show day view call, should I do `insertDB()` as in show calendar view? Looks like it will break the reading stats if a new record is not started when I close it and return to the book I'm reading.

This is what I read as the issue. I thought then that the current page was flushed when opening day view, and a new one was made when closing - so a gap.
But you said the current page is not flushed.
So, I'm not sure I understand your above text.

**Comment by weijiuqiao** [comment]
>This is what I read as the issue. I thought then...

I thought so, too. I thought there would be an issue but it turned out there was no issue :)

**Comment by poire-z** [comment]
Any thought about getting in the timeline vertical thinner/dotted lines to mark every 10mn or 15mn ?

**Comment by weijiuqiao** [comment]
>Any thought about getting in the timeline vertical thinner/dotted lines to mark every 10mn or 15mn ?

Not particularly :) One should be able to tell roughly?

**Comment by poire-z** [comment]
> One should be able to tell roughly?

Yes, "roughly", but hard to guess if a start segment is at 13h20 or 13h40.
I might be using it too much as it is new, but it's useful to guess how long I've read, when was the last time I moved my butt up to check my computer for messages from work - and how many minutes ago was it - and ensure some reasonable frequency so that nobody will notice it :)

**Comment by weijiuqiao** [comment]
>And an arrow below the current minute if showing the current day would be nice too :)

Actually I'm quite preoccupied this week, and maybe next too :/ But I wouldn't mind at all if you could help with the 10/15 min lines and the triangles :)

**Review by poire-z — APPROVED** [review]

**Comment by yparitcher** [comment]
off topic,
I have a Question for the SQL wizards ;)

Is there a query to get the average time per page for a given time range (last month etc), for documents in a specific folder?

**Comment by weijiuqiao** [comment]
I quickly checked that there's no directory info in statistics.sqlite3. While there is directory in bookinfo_cache.sqlite3, there's no md5, so you can only rely on title and authors to join the tables which might not be 100% accurate.
Some like this and then divide the duration by count (not tested):
```
ATTACH path/to/bookinfo_cache AS info_cache;
SELECT count(DISTINCT page, id_book),
           sum(duration)
    FROM   page_stat
    INNER JOIN book b ON id_book = b.id
    INNER JOIN info_cache.bookinfo c ON (c.title, c.authors) = (b.title, b.authors)
    WHERE  start_time BETWEEN ? AND ?
            AND c.directory LIKE 'specific/folder%';
DETACH info_cache;
```
